# Day 3 - Dataset Profiling

## Objectives
- Determine the number of rows and columns.
- Inspect the data types of all columns.
- Identify missing values in each column.
- Check for duplicate records using the `norad_id` column.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("../data/raw/current_catalog.csv")
df1 = df.copy()

## 1. Dataset Shape

The shape of the dataset helps understand its size by displaying the total number of rows and columns.

In [2]:
rows, columns = df1.shape

print("Number of Rows :", rows)
print("Number of Columns :", columns)

Number of Rows : 16649
Number of Columns : 19


## 2. Data Types

Inspecting data types ensures each column is stored in the correct format for analysis.

In [3]:
df1.dtypes

norad_id                     int64
name                        object
object_type                 object
satellite_constellation     object
altitude_km                float64
altitude_category           object
orbital_band                object
congestion_risk             object
inclination                float64
eccentricity               float64
launch_year_estimate         int64
days_in_orbit_estimate       int64
orbit_lifetime_category     object
mean_motion                float64
epoch                       object
data_source                 object
snapshot_date               object
country                     object
last_seen                   object
dtype: object

In [4]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16649 entries, 0 to 16648
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   norad_id                 16649 non-null  int64  
 1   name                     16649 non-null  object 
 2   object_type              16647 non-null  object 
 3   satellite_constellation  16649 non-null  object 
 4   altitude_km              16649 non-null  float64
 5   altitude_category        16649 non-null  object 
 6   orbital_band             16649 non-null  object 
 7   congestion_risk          16649 non-null  object 
 8   inclination              16649 non-null  float64
 9   eccentricity             16649 non-null  float64
 10  launch_year_estimate     16649 non-null  int64  
 11  days_in_orbit_estimate   16649 non-null  int64  
 12  orbit_lifetime_category  16649 non-null  object 
 13  mean_motion              16649 non-null  float64
 14  epoch                 

## 3. Missing Values

Checking missing values helps identify incomplete records that may require cleaning before analysis.

In [5]:
missing_values = df1.isnull().sum()

print(missing_values)

norad_id                   0
name                       0
object_type                2
satellite_constellation    0
altitude_km                0
altitude_category          0
orbital_band               0
congestion_risk            0
inclination                0
eccentricity               0
launch_year_estimate       0
days_in_orbit_estimate     0
orbit_lifetime_category    0
mean_motion                0
epoch                      0
data_source                0
snapshot_date              0
country                    2
last_seen                  0
dtype: int64


## 4. Duplicate Records

The `norad_id` column is used as a unique identifier for each satellite. Duplicate values may indicate repeated records.

In [6]:
duplicate_count = df1["norad_id"].duplicated().sum()

print("Duplicate NORAD IDs:", duplicate_count)

Duplicate NORAD IDs: 0


In [8]:
df1[df1["norad_id"].duplicated()]

,norad_id,name,object_type,satellite_constellation,altitude_km,altitude_category,orbital_band,congestion_risk,inclination,eccentricity,launch_year_estimate,days_in_orbit_estimate,orbit_lifetime_category,mean_motion,epoch,data_source,snapshot_date,country,last_seen


## Summary

- Dataset contains 16,649 rows and 19 columns.
- Data types were inspected successfully.
- Only two missing values were identified (`object_type` and `country`).
- Duplicate records were checked using `norad_id`.

In [9]:
df1["object_type"] = df1["object_type"].fillna("UNKNOWN")
df1["country"] = df1["country"].fillna("UNKNOWN")

In [10]:
missing_values = df1.isnull().sum()

print(missing_values)

norad_id                   0
name                       0
object_type                0
satellite_constellation    0
altitude_km                0
altitude_category          0
orbital_band               0
congestion_risk            0
inclination                0
eccentricity               0
launch_year_estimate       0
days_in_orbit_estimate     0
orbit_lifetime_category    0
mean_motion                0
epoch                      0
data_source                0
snapshot_date              0
country                    0
last_seen                  0
dtype: int64


### Why use `UNKNOWN` instead of `mode()`?

Two missing values were found in the dataset:

- `object_type` → 2 missing values
- `country` → 2 missing values

Although these missing values can be filled using the **mode** (most frequent value), doing so may introduce incorrect information. For example, filling a missing `country` with `USA` or a missing `object_type` with `PAYLOAD` assumes the missing value belongs to the most common category, which cannot be verified.

Instead, these missing values will be replaced with **`UNKNOWN`** during the data cleaning phase. This approach:

- Preserves all records in the dataset.
- Avoids introducing incorrect or misleading information.
- Clearly indicates that the original value was unavailable.
- Maintains better data quality and transparency.

In [13]:
df1[df1.isnull().any(axis=1)].index

Index([], dtype='int64')

In [14]:
missing_rows = df1[df1.isnull().any(axis=1)]
missing_rows


,norad_id,name,object_type,satellite_constellation,altitude_km,altitude_category,orbital_band,congestion_risk,inclination,eccentricity,launch_year_estimate,days_in_orbit_estimate,orbit_lifetime_category,mean_motion,epoch,data_source,snapshot_date,country,last_seen


In [15]:
df1[df1.isnull().any(axis=1)].index

Index([], dtype='int64')

In [16]:
missing_rows = df1[df1.isnull().any(axis=1)]
missing_rows

,norad_id,name,object_type,satellite_constellation,altitude_km,altitude_category,orbital_band,congestion_risk,inclination,eccentricity,launch_year_estimate,days_in_orbit_estimate,orbit_lifetime_category,mean_motion,epoch,data_source,snapshot_date,country,last_seen


In [17]:
# Missing values by column
print(df1.isnull().sum())

# Rows containing missing values
missing_rows = df1[df1.isnull().any(axis=1)]

print("Number of rows with missing values:", len(missing_rows))

missing_rows

norad_id                   0
name                       0
object_type                0
satellite_constellation    0
altitude_km                0
altitude_category          0
orbital_band               0
congestion_risk            0
inclination                0
eccentricity               0
launch_year_estimate       0
days_in_orbit_estimate     0
orbit_lifetime_category    0
mean_motion                0
epoch                      0
data_source                0
snapshot_date              0
country                    0
last_seen                  0
dtype: int64
Number of rows with missing values: 0


,norad_id,name,object_type,satellite_constellation,altitude_km,altitude_category,orbital_band,congestion_risk,inclination,eccentricity,launch_year_estimate,days_in_orbit_estimate,orbit_lifetime_category,mean_motion,epoch,data_source,snapshot_date,country,last_seen
